src/shared/sleep-config.c

```diff
@@ -65,64 +80,36 @@ int parse_sleep_config(const char *verb, bool *ret_allow, char ***ret_modes, cha
-                else
-                        states = strv_new("mem", "standby", "freeze");
-
-        } else if (streq(verb, "hibernate")) {
-                allow = allow_hibernate != 0;
-
-                if (hibernate_mode)
-                        modes = TAKE_PTR(hibernate_mode);
-                else
-                        modes = strv_new("platform", "shutdown");
-
-                if (hibernate_state)
-                        states = TAKE_PTR(hibernate_state);
-                else
-                        states = strv_new("disk");
-
-        } else if (streq(verb, "hybrid-sleep")) {
-                allow = allow_hybrid_sleep > 0 ||
-                        (allow_suspend != 0 && allow_hibernate != 0);
-
-                if (hybrid_mode)
-                        modes = TAKE_PTR(hybrid_mode);
-                else
-                        modes = strv_new("suspend", "platform", "shutdown");
-
-                if (hybrid_state)
-                        states = TAKE_PTR(hybrid_state);
-                else
-                        states = strv_new("disk");
-
-        } else if (streq(verb, "suspend-then-hibernate")) {
-                allow = allow_s2h > 0 ||
-                        (allow_suspend != 0 && allow_hibernate != 0);
-
-                modes = states = NULL;
-        } else
-                assert_not_reached("what verb");
-
-        if ((!modes && STR_IN_SET(verb, "hibernate", "hybrid-sleep")) ||
-            (!states && !streq(verb, "suspend-then-hibernate")))
-                return log_oom();
-
-        if (ret_allow)
-                *ret_allow = allow;
-        if (ret_modes)
-                *ret_modes = TAKE_PTR(modes);
-        if (ret_states)
-                *ret_states = TAKE_PTR(states);
-        if (ret_delay)
-                *ret_delay = delay;
+        /* use default values unless set */
+        sc->allow_suspend = allow_suspend != 0;
+        sc->allow_hibernate = allow_hibernate != 0;
+        sc->allow_hybrid_sleep = allow_hybrid_sleep > 0
+                || (allow_suspend != 0 && allow_hibernate != 0);
+        sc->allow_s2h = allow_s2h > 0
+                || (allow_suspend != 0 && allow_hibernate != 0);
+
+        if (!sc->suspend_states)
+                sc->suspend_states = strv_new("mem", "standby", "freeze");
+        if (!sc->hibernate_modes)
+                sc->hibernate_modes = strv_new("platform", "shutdown");
+        if (!sc->hibernate_states)
+                sc->hibernate_states = strv_new("disk");
+        if (!sc->hybrid_modes)
+                sc->hybrid_modes = strv_new("suspend", "platform", "shutdown");
+        if (!sc->hybrid_states)
+                sc->hybrid_states = strv_new("disk");
+        if (sc->hibernate_delay_sec == 0)
+                sc->hibernate_delay_sec = 180 * USEC_PER_MINUTE;
+
+        /* ensure values set for all required fields */
+        char *v;
+        FOREACH_STRING(v, *sc->suspend_states,
```

Comment:
No, this will not work. FOREACH_STRING iterates until the first NULL.

Just say
```C
if (!sc->suspend_states || !sc->hibernate_modes || !sc->hibernate_states || !sc->hybrid_modes || !sc->hybrid_states)
        return log_oom();
```
There's just 6 of those fields, so no need to overcomplicate things.